packages/table/src/table.tsx

```diff
@@ -586,6 +586,72 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
+        primaryAxis: "row" | "col",
+        secondaryAxis: "row" | "col",
+        isUpOrLeft: boolean,
+        newFocusedCell: IFocusedCellCoordinates,
+        focusCellRegion: IRegion,
+    ) {
+
+        const { grid } = this;
+        const { selectedRegions } = this.state;
+
+        const primaryAxisPlural = primaryAxis === "row" ? "rows" : "cols";
+        const secondaryAxisPlural = secondaryAxis === "row" ? "rows" : "cols";
+
+        const movementDirection = isUpOrLeft ? -1 : +1;
+        const regionIntervalIndex = isUpOrLeft ? 1 : 0;
+
+        // try moving the cell in the direction along the primary axis
+        newFocusedCell[primaryAxis] += movementDirection;
+
+        const isPrimaryIndexOutOfBounds = isUpOrLeft
+            ? newFocusedCell[primaryAxis] < focusCellRegion[primaryAxisPlural][0]
+            : newFocusedCell[primaryAxis] > focusCellRegion[primaryAxisPlural][1];
+
+        if (isPrimaryIndexOutOfBounds) {
+            // if we moved outside the bounds of selection region,
+            // move to the start (or end) of the primary axis, and move one along the secondary
+            newFocusedCell[primaryAxis] = focusCellRegion[primaryAxisPlural][regionIntervalIndex];
+            newFocusedCell[secondaryAxis] += movementDirection;
+
+            const isSecondaryIndexOutOfBounds = isUpOrLeft
+                ? newFocusedCell[secondaryAxis] < focusCellRegion[secondaryAxisPlural][0]
+                : newFocusedCell[secondaryAxis] > focusCellRegion[secondaryAxisPlural][1];
+
+            if (isSecondaryIndexOutOfBounds) {
+                // if moving along the secondary also moves us outside
+                // go to the start (or end) of the next (or previous region)
+                // (note that if there's only one region you'll be moving to the opposite corner, which is fine)
+                let newFocusCellSelectionIndex = newFocusedCell.focusSelectionIndex + movementDirection;
+
+                // newFocusCellSelectionIndex should be one more (or less), unless we need to wrap around
+                if (isUpOrLeft
+                    ? newFocusCellSelectionIndex < 0
+                    : newFocusCellSelectionIndex >= selectedRegions.length
+                ) {
+                    newFocusCellSelectionIndex = isUpOrLeft
+                        ? selectedRegions.length - 1
+                        : 0;
+                }
+
+                const newFocusCellRegion = Regions.getCellRegionFromRegion(
+                    selectedRegions[newFocusCellSelectionIndex],
+                    grid.numRows,
+                    grid.numCols,
+                );
+
+                newFocusedCell = {
+                    col: newFocusCellRegion.cols[regionIntervalIndex],
+                    focusSelectionIndex: newFocusCellSelectionIndex,
+                    row: newFocusCellRegion.rows[regionIntervalIndex],
+                };
+            }
+        }
+        return newFocusedCell;
+    }
```

Comment:
Not sure we saved much in terms of lines of code, but at least it's not redundant!